llvm/utils/TableGen/DXILEmitter.cpp

```diff
@@ -219,6 +269,20 @@ static std::string getOverloadKindStr(MVT::SimpleValueType VT) {
     return "OverloadKind::I16 | OverloadKind::I32 | OverloadKind::I64";
   case MVT::fAny:
     return "OverloadKind::HALF | OverloadKind::FLOAT | OverloadKind::DOUBLE";
+  case MVT::Other:
+    // Handle DXIL-specific overload types
+    {
+      auto RetStr =
+          StringSwitch<std::string>(R->getNameInitAsString())
+              .Case("llvm_i16ori32_ty", "OverloadKind::I16 | OverloadKind::I32")
+              .Case("llvm_halforfloat_ty",
+                    "OverloadKind::HALF | OverloadKind::FLOAT")
```

Comment:
String matching on these two types doesn't feel very robust or scalable - I think we want to do this slightly differently.

One option that keeps this simple but encodes the information a bit more clearly would be to add bits to LLVMType, like so:
```
let isHalfOrFloat = 0
```

This still isn't very extensible but it accomplishes the approach you have here a little bit more cleanly.

The other option is of course to support lists of types and embed them in some kind of LLVMAlternativeType or something, but that's quite a bit more complex and might be overkill.

Reply:
> String matching on these two types doesn't feel very robust or scalable - I think we want to do this slightly differently.
> 
> One option that keeps this simple but encodes the information a bit more clearly would be to add bits to LLVMType, like so:
> 
> ```
> let isHalfOrFloat = 0
> ```
> 
> This still isn't very extensible but it accomplishes the approach you have here a little bit more cleanly.

OK. Changed.

> The other option is of course to support lists of types and embed them in some kind of LLVMAlternativeType or something, but that's quite a bit more complex and might be overkill.

I had considered something like this but refrained for the reasons you cited.

Reply:
Hm, I'm not sure that really worked out that well. We're still switching over the names in "getParameterKind" so we haven't really improved the abstraction, and the if chain over the known set of bools is probably just as error prone.

Reply:
Are these the only two instances of these sets of types that are needed for all of the DXIL ops? If there are others, are they always two types or could it be a larger set?

Reply:
> Hm, I'm not sure that really worked out that well. We're still switching over the names in "getParameterKind" so we haven't really improved the abstraction, and the if chain over the known set of bools is probably just as error prone.

I missed changing this usage. I pushed a change. Thanks!

> Are these the only two instances of these sets of types that are needed for all of the DXIL ops? 

No. It appears that there are other instances of valid overload types of some DXIL Ops that would be over-specified using LLVM's `*Any` value type. For example, valid overload types of [`WaveMatrix_Fill`](https://github.com/microsoft/DirectXShaderCompiler/blob/130877392c263888ef06bab768856d3dab1f1c9a/utils/hct/hctdb.py#L5036) are `half`, `float` and `int`. However, I have not yet formulated how (if?) such DXIL ops should be specified in `DXIL.td`. 

> If there are others, are they always two types or could it be a larger set?

Such overload types can include more than two fixed types.